File: Classes/TableViewController.m
/*

  Created by David Spooner on 5/10/10.
  Copyright 2010 Lambda Software Corporation. All rights reserved.

*/

#import "TableViewController.h"
#import "ColorView.h"


@implementation MyTableViewController

- (id) initWithNibName:(NSString *)aName bundle:(NSBundle *)aBundle
  {
    if ((self = [super initWithNibName:aName bundle:aBundle]) == nil)
      return nil;

    colors = [[NSMutableArray alloc] initWithObjects:
        [UIColor redColor],
        [UIColor greenColor],
        [UIColor blueColor],
        nil];

    return self;
  }


- (void) dealloc
  {
    [colors release];
    [colorPickerViewController release];
    [super dealloc];
  }


- (void) addAction:(id)sender
  {
    UITableView *tableView = self.tableView;
    [tableView beginUpdates];
    NSUInteger index = [colors count];
    [colors addObject:[UIColor lightGrayColor]];
    [tableView insertRowsAtIndexPaths:[NSArray arrayWithObject:[NSIndexPath indexPathForRow:index inSection:0]] withRowAnimation:UITableViewRowAnimationBottom];
    [tableView endUpdates];
  }


#pragma mark NSKeyValueObserving

- (void) observeValueForKeyPath:(NSString *)path ofObject:(id)object change:(NSDictionary *)change context:(void *)context
  {
    NSAssert(object == colorPickerViewController && [path isEqualToString:@"selectedColor"], @"unexpected observation");

    if (selectedRowIndex != NSNotFound) {
      [colors replaceObjectAtIndex:selectedRowIndex withObject:colorPickerViewController.selectedColor];
      [self.tableView reloadRowsAtIndexPaths:[NSArray arrayWithObject:[NSIndexPath indexPathForRow:selectedRowIndex inSection:0]] withRowAnimation:UITableViewRowAnimationFade];
    }
  }


#pragma mark UIViewController

- (void) viewDidLoad
  {
    [super viewDidLoad];

    self.navigationItem.title = @"Colors";

    // Make the right navigation button the Edit button
    self.navigationItem.rightBarButtonItem = self.editButtonItem;

    // Add a toolbar with a right-justified button for taking a picture
    self.toolbarItems = [NSArray arrayWithObjects:
          [[[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemFlexibleSpace target:nil action:nil] autorelease],
          [[[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemAdd target:self action:@selector(addAction:)] autorelease],
          nil];

    UITableView *tableView = self.tableView;
    tableView.dataSource = self;
    tableView.delegate = self;

    colorPickerViewController = [[GxColorPickerViewController alloc] init];
    [colorPickerViewController addObserver:self forKeyPath:@"selectedColor" options:0 context:NULL];

    selectedRowIndex = NSNotFound;
  }


- (void) viewDidUnload
  {
    [super viewDidUnload];

    [colorPickerViewController removeObserver:self forKeyPath:@"selectedColor"];
    [colorPickerViewController release];
    colorPickerViewController = nil;

    selectedRowIndex = NSNotFound;
  }


#pragma mark UITableViewDataSource

- (NSInteger) tableView:(UITableView *)sender numberOfRowsInSection:(NSInteger)section
  { return [colors count]; }


- (UITableViewCell *) tableView:(UITableView *)sender cellForRowAtIndexPath:(NSIndexPath *)path
  {
    UITableViewCell *cell = [sender dequeueReusableCellWithIdentifier:@"Cell"];

    if (cell == nil) {
      cell = [[[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"Cell"] autorelease];
      cell.backgroundView = [[[GxColorView alloc] init] autorelease];
    }

    ((GxColorView *)cell.backgroundView).fillColor = [colors objectAtIndex:path.row];

    return cell;
  }


- (BOOL) tableView:(UITableView *)sender canEditRowAtIndexPath:(NSIndexPath *)path
  { return YES; }


- (void) tableView:(UITableView *)sender commitEditingStyle:(UITableViewCellEditingStyle)style forRowAtIndexPath:(NSIndexPath *)path
  {
    switch (style) {
      case UITableViewCellEditingStyleNone :
        break;
      case UITableViewCellEditingStyleDelete :
        [colors removeObjectAtIndex:path.row];
        [sender deleteRowsAtIndexPaths:[NSArray arrayWithObject:path] withRowAnimation:UITableViewRowAnimationFade];
        break;
      default :
        NSAssert(0, @"unhandled case");
    }
  }


- (BOOL) tableView:(UITableView *)sender canMoveRowAtIndexPath:(NSIndexPath *)path
  { return NO; }


#pragma mark UITableViewDelegate

- (void) tableView:(UITableView *)sender didSelectRowAtIndexPath:(NSIndexPath *)path
  {
    selectedRowIndex = path.row;

    colorPickerViewController.selectedColor = [colors objectAtIndex:selectedRowIndex];

    [self.navigationController pushViewController:colorPickerViewController animated:YES];
  }

@end

File: ColorWheel-iOS/TableViewController.swift
/*

  Copyright © 2010-2016 David Spooner; see License.txt

  A UITableViewController subclass to present a list of colors for inspection by ColorPickerViewController.

*/

import UIKit


class ColorWheelTableViewController : UITableViewController
  {

    var colorPickerViewController: ColorPickerViewController?
    var colors: [UIColor] = []
    var selectedRow: NSInteger = 0


    convenience init()
      {
        self.init(style: UITableViewStyle.plain)

        colors = [ UIColor.red, UIColor.green, UIColor.blue ]

        colorPickerViewController = ColorPickerViewController()
        colorPickerViewController!.addObserver(self, forKeyPath:"selectedColor", options:NSKeyValueObservingOptions(), context:nil)

        self.title = "Colors"
      }


    deinit
      {
        colorPickerViewController?.removeObserver(self, forKeyPath:"selectedColor")
      }


    func add(_ sender: UIBarButtonItem)
      {
        // Add a new element to the end of the list
        colors.append(UIColor.lightGray)

        // Inform the table view of the row insertion
        self.tableView.insertRows(at: [IndexPath(row:colors.count-1, section:0)], with:UITableViewRowAnimation.fade);
      }


    // NSKeyValueObserving

    override func observeValue(forKeyPath path: String?, of sender: Any?, change: [NSKeyValueChangeKey:Any]?, context:UnsafeMutableRawPointer?)
      {
        // Update the corresponding list element
        colors[selectedRow] = colorPickerViewController!.selectedColor

        // Ask the table view to reload the effected row
        self.tableView.reloadRows(at: [IndexPath(row:selectedRow, section:0)], with:UITableViewRowAnimation.fade)
      }


    // UIViewController

    override func viewDidLoad()
      {
        super.viewDidLoad()

        self.edgesForExtendedLayout = UIRectEdge()

        self.navigationItem.rightBarButtonItem = UIBarButtonItem(barButtonSystemItem:UIBarButtonSystemItem.add, target:self, action:#selector(ColorWheelTableViewController.add(_:)))
      }


    // UITableViewDataSource

    override func tableView(_ sender: UITableView, numberOfRowsInSection section: Int) -> Int
      {
        return colors.count
      }


    override func tableView(_ sender: UITableView, cellForRowAt path: IndexPath) -> UITableViewCell
      {
        var cell = sender.dequeueReusableCell(withIdentifier: "cell")
        if cell == nil {
          cell = UITableViewCell(style:UITableViewCellStyle.default, reuseIdentifier:"cell")
          cell!.backgroundView = ColorView()
        }

        (cell!.backgroundView as! ColorView).fillColor = colors[path.row] as UIColor

        return cell!
      }


    override func tableView(_ sender: UITableView, commit style: UITableViewCellEditingStyle, forRowAt path: IndexPath)
      {
        assert(style == UITableViewCellEditingStyle.delete, "unexpected argument")

        // Remove the indicated list element
        colors.remove(at: path.row);

        // Inform the table view of the row removal
        sender.deleteRows(at: [path], with:UITableViewRowAnimation.fade);
      }


    // UITableViewDelegate

    override func tableView(_ sender: UITableView, didSelectRowAt path: IndexPath)
      {
        // Retain the selected row for posterity
        selectedRow = path.row;

        // Set the current color of the color picker
        colorPickerViewController!.selectedColor = colors[selectedRow]

        // Navigate to the color picker view
        self.navigationController?.pushViewController(colorPickerViewController!, animated:true)
      }

  }
